### PR TITLE
Add Job for deleting tenants

### PR DESCRIPTION
**Pull request by welzfabian**

### CODE REVIEW

File: src/main/java/sirius/biz/tenants/deletion/DeleteSchedulerEntriesWorkItem.java [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+    private SchedulerEntryProvider schedulerEntryProvider;
+
+    @Part
+    private Mixing mixing;
+
+    @Override
+    public <I, T extends BaseEntity<I> & Tenant<I>> ProcessLog execute(T tenant, boolean simulate) {
+        BaseMapper mapper = mixing.getDescriptor(schedulerEntryProvider.getSchedulerEntryType()).getMapper();
+
+        long schedulerEntryCount = ((Query) mapper.select(schedulerEntryProvider.getSchedulerEntryType())
```
Comment (sabieber) [inline src/main/java/sirius/biz/tenants/deletion/DeleteSchedulerEntriesWorkItem.java]: Could the queries be extracted? They are built two times for every workitem

File: src/main/java/sirius/biz/tenants/deletion/TenantDeletionWorkItem.java [hunk 2]
```diff
@@ -13,10 +13,29 @@
 import sirius.db.mixing.BaseEntity;
 import sirius.kernel.di.std.Priorized;
 
+/**
+ * Objects of this interface will be called when a tenant is deleted and are responsible for deleting external tenant
+ * data, e.g {@link sirius.biz.tenants.UserAccount UserAccounts} in {@link DeleteUserAccountsWorkItem}.
```
Comment (sabieber) [inline src/main/java/sirius/biz/tenants/deletion/TenantDeletionWorkItem.java]: Hint that classes implementing this interface must be registered?

File: src/main/java/sirius/biz/tenants/deletion/DeleteTenantJob.java [hunk 3]
```diff
@@ -145,5 +106,11 @@ public String getCategory() {
         public String getIcon() {
             return "fa-trash";
         }
+
+        @Nonnull
+        @Override
+        public String getName() {
+            return "delete-tenant";
```
Comment (sabieber) [inline src/main/java/sirius/biz/tenants/deletion/DeleteTenantJob.java]: Do the controllers still link to the old specific jobs?

File: src/main/java/sirius/biz/tenants/deletion/DeleteCodeListsWorkItem.java [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+import sirius.db.mixing.Mixing;
+import sirius.db.mixing.query.Query;
+import sirius.kernel.di.std.Part;
+import sirius.kernel.di.std.Register;
+
+/**
+ * Deletes the {@link CodeList codeLists} and {@link CodeListEntry codeListEntries} of the given tenant.
+ */
+@Register
+public class DeleteCodeListsWorkItem implements TenantDeletionWorkItem {
```
Comment (andyHa) [inline src/main/java/sirius/biz/tenants/deletion/DeleteCodeListsWorkItem.java]: wouldn't this code be much more readable if we'd provide a concrete impl for mongo & sql?

File: src/main/java/sirius/biz/tenants/deletion/DeleteSchedulerEntriesWorkItem.java [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+import sirius.db.mixing.Mixing;
+import sirius.db.mixing.query.Query;
+import sirius.kernel.di.std.Part;
+import sirius.kernel.di.std.Register;
+
+/**
+ * Deletes {@link sirius.biz.jobs.scheduler.SchedulerEntry schedulerEntries} of the given tenant.
+ */
+@Register
+public class DeleteSchedulerEntriesWorkItem implements TenantDeletionWorkItem {
```
Comment (andyHa) [inline src/main/java/sirius/biz/tenants/deletion/DeleteSchedulerEntriesWorkItem.java]: s.a.

File: src/main/java/sirius/biz/tenants/deletion/DeleteTenantJob.java [hunk 6]
```diff
@@ -0,0 +1,116 @@
...
+
+import javax.annotation.Nonnull;
+import java.util.List;
+import java.util.Map;
+import java.util.function.Consumer;
+
+/**
+ * Deletes a given tenant and the related data.
+ * <p>
+ * Implement and register a {@link TenantDeletionWorkItem} object to implement logic for deleting external tenant data,
```
Comment (andyHa) [inline src/main/java/sirius/biz/tenants/deletion/DeleteTenantJob.java]:
```suggestion
 * Implement and register a {@link TenantDeletionWorkItem} to implement logic for deleting external tenant data,
```

File: src/main/java/sirius/biz/tenants/deletion/DeleteTenantJob.java [hunk 7]
```diff
@@ -0,0 +1,116 @@
...
+        }
+
+        for (TenantDeletionWorkItem deletionWorkItem : deletionWorkItems) {
+            process.log(deletionWorkItem.execute(tenant, simulate));
+        }
+    }
+
+    @Register(classes = JobFactory.class, framework = Tenants.FRAMEWORK_TENANTS)
+    @Permission(TenantController.PERMISSION_MANAGE_TENANTS)
+    public static class DeleteTenantJobFactory extends DefaultBatchProcessFactory {
```
Comment (andyHa) [inline src/main/java/sirius/biz/tenants/deletion/DeleteTenantJob.java]: SimpleBatchProcessFactory ?

File: src/main/java/sirius/biz/tenants/deletion/DeleteTenantJob.java [hunk 8]
```diff
@@ -0,0 +1,116 @@
...
+
+        protected static final TenantParameter TENANT_PARAMETER =
+                new TenantParameter("tenant", "$Model.tenant").markRequired();
+
+        protected static final BooleanParameter SIMULATION_PARAMETER =
+                new BooleanParameter("simulate", "$DeleteTenantJobFactory.simulate");
+
+        @Part
+        protected Tenants<?, ?, ?> tenants;
```
Comment (andyHa) [inline src/main/java/sirius/biz/tenants/deletion/DeleteTenantJob.java]: redundant?

File: src/main/java/sirius/biz/tenants/deletion/DeleteTenantJob.java [hunk 9]
```diff
@@ -0,0 +1,116 @@
...
+
+        @Override
+        protected String createProcessTitle(Map<String, String> context) {
+            return NLS.fmtr("DeleteTenantJobFactory.title")
+                      .set("name", tenants.fetchCachedTenantName(context.get("tenant")))
```
Comment (andyHa) [inline src/main/java/sirius/biz/tenants/deletion/DeleteTenantJob.java]: use the parameter to resolve the id into a tenant - this assums the implementation of the parameter which isn't guaranteed

File: src/main/java/sirius/biz/tenants/deletion/DeleteTenantJob.java [hunk 10]
```diff
@@ -0,0 +1,116 @@
...
+import java.util.Map;
+import java.util.function.Consumer;
+
+/**
+ * Deletes a given tenant and the related data.
+ * <p>
+ * Implement and register a {@link TenantDeletionWorkItem} object to implement logic for deleting external tenant data,
+ * like {@link UserAccount UserAccounts} for example.
+ */
+public class DeleteTenantJob extends BatchJob {
```
Comment (andyHa) [inline src/main/java/sirius/biz/tenants/deletion/DeleteTenantJob.java]: permission?!

File: src/main/java/sirius/biz/tenants/deletion/DeleteTenantWorkItem.java [hunk 11]
```diff
@@ -0,0 +1,43 @@
...
+import sirius.db.mixing.BaseMapper;
+import sirius.db.mixing.Mixing;
+import sirius.kernel.di.std.Part;
+import sirius.kernel.di.std.Register;
+
+/**
+ * Deletes the given tenant.
+ */
+@Register
+public class DeleteTenantWorkItem implements TenantDeletionWorkItem {
```
Comment (andyHa) [inline src/main/java/sirius/biz/tenants/deletion/DeleteTenantWorkItem.java]: overkill? could be the last step in the job...

File: src/main/java/sirius/biz/tenants/deletion/DeleteUserAccountsWorkItem.java [hunk 12]
```diff
@@ -0,0 +1,50 @@
...
+import sirius.db.mixing.Mixing;
+import sirius.db.mixing.query.Query;
+import sirius.kernel.di.std.Part;
+import sirius.kernel.di.std.Register;
+
+/**
+ * Deletes all {@link UserAccount userAccounts} of the given tenant.
+ */
+@Register
+public class DeleteUserAccountsWorkItem implements TenantDeletionWorkItem {
```
Comment (andyHa) [inline src/main/java/sirius/biz/tenants/deletion/DeleteUserAccountsWorkItem.java]: s.a.

File: src/main/java/sirius/biz/tenants/deletion/DeleteUserAccountsWorkItem.java [hunk 13]
```diff
@@ -0,0 +1,50 @@
...
+    public ProcessLog execute(Tenant<?> tenant, boolean simulate) {
+        BaseMapper mapper = mixing.getDescriptor(tenants.getUserClass()).getMapper();
+
+        long userCount = ((Query) mapper.select(tenants.getUserClass()).eq(UserAccount.TENANT, tenant)).count();
+
+        if (!simulate) {
+            ((Query) mapper.select(tenants.getUserClass()).eq(UserAccount.TENANT, tenant)).delete();
+        }
+
+        return ProcessLog.info().withNLSKey(getMessageKey(simulate)).withContext("userCount", Long.toString(userCount));
```
Comment (andyHa) [inline src/main/java/sirius/biz/tenants/deletion/DeleteUserAccountsWorkItem.java]: NLS.toUserString

File: src/main/java/sirius/biz/tenants/deletion/TenantDeletionWorkItem.java [hunk 14]
```diff
@@ -0,0 +1,43 @@
...
+ */
+
+package sirius.biz.tenants.deletion;
+
+import sirius.biz.process.logs.ProcessLog;
+import sirius.biz.tenants.Tenant;
+import sirius.kernel.di.std.Priorized;
+
+/**
+ * Objects of this interface will be called when a tenant is deleted and are responsible for deleting external tenant
```
Comment (andyHa) [inline src/main/java/sirius/biz/tenants/deletion/TenantDeletionWorkItem.java]:
```suggestion
 * Classes implementing this interface will be called when a tenant is deleted and are responsible for deleting 
```

File: src/main/java/sirius/biz/tenants/deletion/TenantDeletionWorkItem.java [hunk 15]
```diff
@@ -0,0 +1,43 @@
...
+public interface TenantDeletionWorkItem extends Priorized {
+
+    /**
+     * Called with the {@link Tenant} for which data should be deleted.
+     *
+     * @param tenant   the tenant which is deleted
+     * @param simulate wheter the deleting should only be simulated
+     * @return A {@link ProcessLog} message, whichis logged by the executing job
+     */
+    ProcessLog execute(Tenant<?> tenant, boolean simulate);
```
Comment (andyHa) [inline src/main/java/sirius/biz/tenants/deletion/TenantDeletionWorkItem.java]: pass the ProcessContext in to provide more flexibility..

File: src/main/java/sirius/biz/tenants/deletion/TenantDeletionWorkItem.java [hunk 16]
```diff
@@ -0,0 +1,43 @@
...
+
+    /**
+     * Returns the message key for the {@link ProcessLog} returned by the {@link #execute(Tenant, boolean)} method.
+     *
+     * @param simulate wheter the deleting should only be simulated
+     * @return the message key for the log message
+     */
+    default String getMessageKey(boolean simulate) {
```
Comment (andyHa) [inline src/main/java/sirius/biz/tenants/deletion/TenantDeletionWorkItem.java]: no default methods in new code - this is a Java Hack to enhance old interfaces..

File: src/main/java/sirius/biz/tenants/jdbc/SQLTenantController.java [hunk 17]
```diff
@@ -47,6 +48,11 @@
         return pageHelper;
     }
 
+    @Override
+    public String getDeleteLink(String tenantId) {
+        return new LinkBuilder("/job/delete-sql-tenant").append("simulate", true).append("tenant", tenantId).toString();
```
Comment (andyHa) [inline src/main/java/sirius/biz/tenants/jdbc/SQLTenantController.java]: can't the job itself be generic as it doesn't do anything?

File: src/main/resources/biz_de.properties [hunk 18]
```diff
@@ -540,6 +540,19 @@ FieldDefinition.name=Name
 FieldDefinition.type=Typ
 FieldDefinition.remarks=Bemerkungen
 ImportDictionary.automaticMappings=Die Spaltenzuordnungen werden automatisch anhand der ersten Zeile und der Feldnamen (bzw. alternativen Bezeichnungen) ermittelt. Hierbei werden Groß/Kleinschreibung sowie Leerzeichen ignoriert.
+DeleteTenantJob.simulate = Die Löschung wird nur simuliert.
+DeleteTenantJobFactory.simulate = Simulieren
+DeleteTenantJobFactory.label = Mandant löschen
+DeleteTenantJobFactory.description = Löscht einen ausgewählten Mandanten und alle zugehörigen Daten.
+DeleteTenantJobFactory.title = Lösche Mandanten "${name}"
+DeleteCodeListsWorkItem.message = Es wurden ${codeListCount} Code-Listen gelöscht.
+DeleteCodeListsWorkItem.message.simulate = SIMULIERT: Es wurden ${codeListCount} Code-Listen gelöscht.
```
Comment (andyHa) [inline src/main/resources/biz_de.properties]: couldn't we do this in a generic way?

Comment (andyHa) [reply src/main/resources/biz_de.properties]: (always append a common suffix like (simulated))

File: src/main/java/sirius/biz/codelists/CodeLists.java [hunk 19]
```diff
@@ -77,14 +77,14 @@
      *
      * @return the effective implementation of {@link CodeList}
      */
-    protected abstract Class<L> getListType();
+    public abstract Class<L> getListType();
```
Comment (andyHa) [inline src/main/java/sirius/biz/codelists/CodeLists.java]: reset changes (also all below)

File: src/main/java/sirius/biz/jobs/scheduler/SchedulerEntryProvider.java [hunk 20]
```diff
@@ -43,4 +43,11 @@
      * @param timestamp the timestamp when the execution was scheduled
      */
     void markExecuted(J job, LocalDateTime timestamp);
+
+    /**
+     * Returns the type of the handled {@link SchedulerEntry}.
+     *
+     * @return the {@link Class} of the handled {@link SchedulerEntry}
+     */
+    Class<J> getSchedulerEntryType();
```
Comment (andyHa) [inline src/main/java/sirius/biz/jobs/scheduler/SchedulerEntryProvider.java]: s.a.

File: src/main/java/sirius/biz/jobs/scheduler/jdbc/SQLSchedulerEntryProvider.java [hunk 21]
```diff
@@ -57,4 +57,9 @@ public void markExecuted(SQLSchedulerEntry job, LocalDateTime timestamp) {
         job.getSchedulerData().rememberExecution(timestamp);
         oma.update(job);
     }
+
+    @Override
```
Comment (andyHa) [inline src/main/java/sirius/biz/jobs/scheduler/jdbc/SQLSchedulerEntryProvider.java]: s.a.

File: src/main/java/sirius/biz/jobs/scheduler/mongo/MongoSchedulerEntryProvider.java [hunk 22]
```diff
@@ -60,4 +60,9 @@ public void markExecuted(MongoSchedulerEntry job, LocalDateTime timestamp) {
         job.getSchedulerData().rememberExecution(timestamp);
         mango.update(job);
     }
+
+    @Override
```
Comment (andyHa) [inline src/main/java/sirius/biz/jobs/scheduler/mongo/MongoSchedulerEntryProvider.java]: s.a.

File: src/main/java/sirius/biz/codelists/jdbc/SQLCodeLists.java [hunk 23]
```diff
@@ -23,12 +23,12 @@
     public static final String FRAMEWORK_CODE_LISTS_JDBC = "biz.code-lists-jdbc";
 
     @Override
-    protected Class<SQLCodeListEntry> getEntryType() {
+    public Class<SQLCodeListEntry> getEntryType() {
```
Comment (andyHa) [inline src/main/java/sirius/biz/codelists/jdbc/SQLCodeLists.java]: s.a.

File: src/main/java/sirius/biz/codelists/mongo/MongoCodeLists.java [hunk 24]
```diff
@@ -23,12 +23,12 @@
     public static final String FRAMEWORK_CODE_LISTS_MONGO = "biz.code-lists-mongo";
 
     @Override
-    protected Class<MongoCodeListEntry> getEntryType() {
+    public Class<MongoCodeListEntry> getEntryType() {
```
Comment (andyHa) [inline src/main/java/sirius/biz/codelists/mongo/MongoCodeLists.java]: s.a.

**Pull request merged**